narwhal/consensus/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,568 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::bullshark::Bullshark;
+use crate::consensus::ConsensusProtocol;
+use crate::consensus::ConsensusState;
+use crate::consensus_utils::make_consensus_store;
+use crate::metrics::ConsensusMetrics;
+use config::{Committee, Stake};
+use crypto::PublicKey;
+use fastcrypto::hash::Hash;
+use fastcrypto::hash::HashFunction;
+use prometheus::Registry;
+use rand::distributions::Bernoulli;
+use rand::distributions::Distribution;
+use rand::prelude::SliceRandom;
+use rand::rngs::StdRng;
+use rand::Rng;
+use rand::SeedableRng;
+use std::collections::{BTreeSet, HashMap, HashSet, VecDeque};
+use std::num::NonZeroUsize;
+use std::ops::RangeInclusive;
+use std::sync::Arc;
+use test_utils::mock_certificate_with_rand;
+use test_utils::CommitteeFixture;
+#[allow(unused_imports)]
+use tokio::sync::mpsc::channel;
+use types::Round;
+use types::{Certificate, CertificateDigest};
+
+#[derive(Copy, Clone, Debug)]
+pub struct FailureModes {
+    // The probability of having failures per round. The failures should
+    // be <=f , otherwise no DAG could be created. The provided number gives the probability of having
+    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% change of
+    // having failures up to 33% of the nodes.
+    pub nodes_failure_probability: f64,
+
+    // The percentage of slow nodes we want to introduce to our sample. Basically a slow node is one
+    // that might be able to produce certificates, but those are never get referenced by others. That has
+    // as an effect that when they are leaders might also not get enough support - or no support at all.
+    // For example, a value of 0.2 means that we want up to 20% of our nodes to behave as slow nodes.
+    pub slow_nodes_percentage: f64,
```

Comment:
Is it possible to make every node to have non-zero probability to not be included as a parent? I think it will give more variance to the DAG. But the slow nodes setting may be more realistic. We may be above to simulate commit delay with this.

Reply:
@mwtian I don't have this set behind a specific probability (although I wanted to - or more specifically chose a distribution that could fit the purpose) but choosing the parents it is also done randomly (uniformly) between a range of `2f + 1  to 3f + 1` . You can see that here https://github.com/MystenLabs/sui/pull/9239/files#diff-8573c7e7cc360f7537b7906716b2e7b99ca41e220054310f9915b47e30bada3dR305

So that would create cases where some parents might not get picked up at all - or picked up by a subset of children certificates (as we currently expect to happen in real life).